### PR TITLE
NumPy Scalar to PyTorch Scalar

### LINKED ISSUE #4985: np.intN types give TypeErrors
- OS: macos 10.11.6
- PyTorch version: 0.3post4
- How you installed PyTorch (conda, pip, source): pip
- Python version: 3.6.3

`torch.Tensor(np.int64(1))` gives a `TypeError` because torch expects ints, and it got the type np.int64. However, that's still an int (for all practical purposes, even if `isinstance(np.int64(1), int)` equals False.

### PR DESCRIPTION
**Pull request by vishwakftw**
Fixes https://github.com/pytorch/pytorch/issues/4985 .

### CODE REVIEW

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -141,6 +141,9 @@ ScalarType infer_scalar_type(PyObject *obj) {
   if (PyArray_Check(obj)) {
     auto array = (PyArrayObject*)obj;
     return numpy_dtype_to_aten(PyArray_TYPE(array));
+  } else if (PyArray_CheckScalar(obj)) {
```
Comment (li-roy) [inline torch/csrc/utils/tensor_new.cpp]: nit: use an if rather than an else if here to match the code around it

Comment (vishwakftw) [reply torch/csrc/utils/tensor_new.cpp]: Sure, thank you for pointing it out.

**Review by li-roy — APPROVED** [review]

File: torch/csrc/utils/tensor_new.cpp [hunk 2]
```diff
@@ -142,6 +142,10 @@ ScalarType infer_scalar_type(PyObject *obj) {
     auto array = (PyArrayObject*)obj;
     return numpy_dtype_to_aten(PyArray_TYPE(array));
   }
+  if (PyArray_CheckScalar(obj)) {
+    auto array = (PyArrayObject*)(PyArray_FromScalar(obj, NULL));
```
Comment (colesbury) [inline torch/csrc/utils/tensor_new.cpp]: Does this need to free `array`?

**Comment by vishwakftw** [comment]
Is this good to go? Sorry about the reminder.

**Comment by ezyang** [comment]
No worries, reminders appreciated

**Comment by vishwakftw** [comment]
Is this good to go?

**Comment by ailzhang** [comment]
cc: @li-roy @soumith

**Pull request closed without merging**